File: crates/pk-core/src/fx/ring_pool.rs
//! Shockwave Ring Pool — 16-slot ring buffer for expanding/collapsing floor shockwaves.
//!
//! PORTS: `fx/pools/ring-pool.ts`

pub const RING_COUNT: usize = 16;
pub const RING_INNER: f32 = 0.78; // fat unit-ring inner radius -> a soft field band
pub const RING_INNER_THIN: f32 = 0.955; // sharp unit-ring inner radius -> a wave-front line

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct RingOpts {
    pub delay: f32,
    pub inward: bool,
    pub thin: bool,
    pub opacity: f32,
}

impl Default for RingOpts {
    fn default() -> Self {
        Self {
            delay: 0.0,
            inward: false,
            thin: false,
            opacity: 1.0,
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct RingInstance {
    pub active: bool,
    pub x: f32,
    pub z: f32,
    pub current_r: f32,
    pub max_r: f32,
    pub life: f32,
    pub max_life: f32,
    pub delay: f32,
    pub inward: bool,
    pub thin: bool,
    pub peak_opacity: f32,
    pub current_opacity: f32,
}

impl Default for RingInstance {
    fn default() -> Self {
        Self {
            active: false,
            x: 0.0,
            z: 0.0,
            current_r: 0.0,
            max_r: 1.0,
            life: 0.0,
            max_life: 0.0,
            delay: 0.0,
            inward: false,
            thin: false,
            peak_opacity: 1.0,
            current_opacity: 0.0,
        }
    }
}

#[derive(Clone, Debug)]
pub struct RingPool {
    pub slots: [RingInstance; RING_COUNT],
    cursor: usize,
}

impl Default for RingPool {
    fn default() -> Self {
        Self::new()
    }
}

impl RingPool {
    pub fn new() -> Self {
        Self {
            slots: [RingInstance::default(); RING_COUNT],
            cursor: 0,
        }
    }

    pub fn spawn(&mut self, x: f32, z: f32, max_r: f32, duration: f32, opts: Option<RingOpts>) {
        let opt = opts.unwrap_or_default();
        let idx = self.cursor;
        self.cursor = (self.cursor + 1) % RING_COUNT;

        let slot = &mut self.slots[idx];
        slot.active = true;
        slot.x = x;
        slot.z = z;
        slot.max_r = max_r.max(0.1);
        slot.max_life = duration.max(0.01);
        slot.life = slot.max_life;
        slot.delay = opt.delay.max(0.0);
        slot.inward = opt.inward;
        slot.thin = opt.thin;
        slot.peak_opacity = opt.opacity.clamp(0.0, 1.0);
        slot.current_r = if opt.inward { slot.max_r } else { 0.0 };
        slot.current_opacity = 0.0;
    }

    pub fn step(&mut self, dt: f32) {
        for slot in &mut self.slots {
            if slot.life <= 0.0 {
                slot.active = false;
                continue;
            }

            if slot.delay > 0.0 {
                slot.delay -= dt;
                slot.current_opacity = 0.0;
                continue;
            }

            slot.life -= dt;
            if slot.life <= 0.0 {
                slot.active = false;
                slot.current_opacity = 0.0;
                continue;
            }

            let t = (1.0 - slot.life / slot.max_life).clamp(0.0, 1.0);
            // Outward: ease-out expansion (1 - (1-t)^2)
            // Inward: ease-in collapse (1 - t^2)
            let r = if slot.inward {
                slot.max_r * (1.0 - t * t)
            } else {
                slot.max_r * (1.0 - (1.0 - t) * (1.0 - t))
            };
            slot.current_r = r.max(0.05);
            slot.current_opacity = (t * std::f32::consts::PI).sin() * slot.peak_opacity;
        }
    }

    pub fn active_count(&self) -> usize {
        self.slots.iter().filter(|s| s.active).count()
    }
}

File: ThreeJS/src/game/pinball-knight/fx/pools/ring-pool.ts
/**
 * Shockwave rings — flat expanding bands on the floor.
 *
 * Extracted from the 1700-line `vfx.ts` when it moved into `fx/`. Behaviour is
 * unchanged — only the file boundary is new, and the shared constants it used to
 * hold privately now live in `./shared.ts` so eight modules cannot drift apart.
 */
import * as THREE from "three";

const RING_COUNT = 16;
const RING_INNER = 0.78; // fat unit-ring inner radius → a soft field band
const RING_INNER_THIN = 0.955; // sharp unit-ring inner radius → a wave-front line

/** Per-spawn ring styling. */
export interface RingOpts {
  /** Hold the ring hidden this long before it starts (chaser rings). */
  delay?: number;
  /** COLLAPSE inward (a pull) instead of expanding (a push). */
  inward?: boolean;
  /** Draw the sharp wave-front line instead of the fat field band. */
  thin?: boolean;
  /** Peak opacity multiplier (default 1) — how loud this ring is allowed to be. */
  opacity?: number;
}

export class RingPool {
  readonly group: THREE.Group;
  private meshes: THREE.Mesh[] = [];
  private life: number[] = [];
  private maxLife: number[] = [];
  private maxR: number[] = [];
  private delay: number[] = [];
  /** true = the ring COLLAPSES inward (a pull) instead of expanding (a push). */
  private inward: boolean[] = [];
  private peak: number[] = [];
  private fat: THREE.RingGeometry;
  private thin: THREE.RingGeometry;
  private cursor = 0;

  constructor() {
    this.group = new THREE.Group();
    this.fat = new THREE.RingGeometry(RING_INNER, 1, 40);
    this.thin = new THREE.RingGeometry(RING_INNER_THIN, 1, 64);
    for (let i = 0; i < RING_COUNT; i++) {
      const mat = new THREE.MeshBasicMaterial({
        transparent: true,
        depthWrite: false,
        depthTest: true,
        blending: THREE.AdditiveBlending,
        side: THREE.DoubleSide,
        opacity: 0,
      });
      const m = new THREE.Mesh(this.fat, mat);
      m.rotation.x = -Math.PI / 2; // lay flat on the floor
      m.visible = false;
      m.renderOrder = 11;
      this.meshes.push(m);
      this.life.push(0);
      this.maxLife.push(0);
      this.maxR.push(1);
      this.delay.push(0);
      this.inward.push(false);
      this.peak.push(1);
      this.group.add(m);
    }
  }

  spawn(x: number, z: number, color: number, maxRadius: number, duration: number, opts: RingOpts = {}): void {
    const i = this.cursor;
    this.cursor = (this.cursor + 1) % RING_COUNT;
    const m = this.meshes[i];
    m.position.set(x, 0.06, z);
    m.geometry = opts.thin ? this.thin : this.fat;
    (m.material as THREE.MeshBasicMaterial).color.setHex(color);
    m.visible = false; // stays hidden through the delay
    this.life[i] = duration;
    this.maxLife[i] = duration;
    this.maxR[i] = maxRadius;
    this.delay[i] = opts.delay ?? 0;
    this.inward[i] = opts.inward ?? false;
    this.peak[i] = opts.opacity ?? 1;
  }

  update(dt: number): void {
    for (let i = 0; i < RING_COUNT; i++) {
      if (this.life[i] <= 0) continue;
      if (this.delay[i] > 0) {
        this.delay[i] -= dt;
        continue;
      }
      this.life[i] -= dt;
      const m = this.meshes[i];
      if (this.life[i] <= 0) {
        m.visible = false;
        continue;
      }
      const t = 1 - this.life[i] / this.maxLife[i]; // 0 → 1
      // Outward: ease-out expansion (a pressure wave — fast launch, slow edge).
      // Inward: ease-IN collapse, so a pull ACCELERATES into the centre, which
      // is what makes it read as suction rather than a wave played backwards.
      const r = this.inward[i] ? this.maxR[i] * (1 - t * t) : this.maxR[i] * (1 - (1 - t) * (1 - t));
      m.scale.setScalar(Math.max(0.05, r));
      (m.material as THREE.MeshBasicMaterial).opacity = Math.sin(t * Math.PI) * this.peak[i];
      m.visible = true;
    }
  }

  /** See `warmupReveal` — slot 0 stands in for the pool at prewarm time. */
  warmupTarget(): THREE.Object3D {
    return this.meshes[0];
  }

  dispose(): void {
    this.fat.dispose();
    this.thin.dispose();
    for (const m of this.meshes) (m.material as THREE.Material).dispose();
  }
}

/**
 * RUNE SIGIL — a summoning glyph that snaps into existence under a cast, spins,
 * and burns away.
 *
 * This is the piece that makes a spell read as MAGIC rather than as geometry.
 * A shockwave ring alone is a circle; a circle with runes, radial ticks and a
 * counter-rotating inner wheel is a spell being cast. Painted once to a canvas
 * (the slash/floor-fx pattern) and tinted per spawn, so it costs one texture.
 *
 * The sigil punches in fast (over-scaled, then settling), holds while the wave
 * travels, and fades as it over-expands — never a static decal.
 */
// Headroom for overlap: a pulse strikes two at once, the magnet re-strikes one
// per beat while its aura runs, and a cast flourish can land on top of both.
